client/src/procedures.rs

```diff
@@ -0,0 +1,440 @@
+            change_log: Vec::new(),
+        };
+        match self.execute(runner, &mut state) {
+            Ok(()) => {
+                // Delete temporary records
+                Self::revoke_records(runner, state.change_log, true);
+                let mut output = HashMap::new();
+                for (k, (data, is_temp)) in state.aggregated_output.into_iter() {
+                    if !is_temp {
+                        output.insert(k, data);
+                    }
+                }
+                Ok(CollectedOutput { output })
+            }
+            Err(e) => {
+                // Rollback written data
+                Self::revoke_records(runner, state.change_log, false);
+                Err(e)
+            }
+        }
+    }
+
+    fn revoke_records<R: Runner>(runner: &mut R, logs: Vec<ChangeLog>, remove_only_temp: bool) {
+        let mut vaults = HashSet::new();
+        for entry in logs {
+            if entry.is_temp || !remove_only_temp {
+                let (v, _) = SecureClient::resolve_location(&entry.location);
+                let _ = runner.revoke_data(&entry.location);
+                vaults.insert(v);
+            }
+        }
+        for vault_id in vaults {
+            let _ = runner.garbage_collect(vault_id);
+        }
+    }
+}
+
+impl ProcedureStep for Procedure {
+    fn execute<R: Runner>(self, runner: &mut R, state: &mut State) -> Result<(), anyhow::Error> {
+        self.inner.into_iter().try_for_each(|p| p.execute(runner, state))
+    }
+}
+
+impl Message for Procedure {
+    type Result = Result<CollectedOutput, anyhow::Error>;
+}
+
+impl<P> From<P> for Procedure
+where
+    P: Into<PrimitiveProcedure>,
+{
+    fn from(p: P) -> Self {
+        Self { inner: vec![p.into()] }
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct State {
+    aggregated_output: HashMap<OutputKey, (ProcedureIo, bool)>,
+    change_log: Vec<ChangeLog>,
+}
+
+impl State {
+    pub fn insert_data(&mut self, key: OutputKey, value: ProcedureIo, is_temp: bool) {
+        self.aggregated_output.insert(key, (value, is_temp));
+    }
+
+    pub fn get_data(&self, key: &OutputKey) -> Result<&ProcedureIo, anyhow::Error> {
+        self.aggregated_output
+            .get(key)
+            .map(|(data, _)| data)
+            .ok_or_else(|| anyhow::anyhow!("Missing Data"))
+    }
+
+    pub fn add_log(&mut self, location: Location, is_temp: bool) {
+        let log = ChangeLog { location, is_temp };
+        self.change_log.push(log)
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ChangeLog {
+    location: Location,
+    is_temp: bool,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct ProcedureIo(Vec<u8>);
+
+impl Deref for ProcedureIo {
+    type Target = Vec<u8>;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl From<Vec<u8>> for ProcedureIo {
+    fn from(v: Vec<u8>) -> Self {
+        ProcedureIo(v)
+    }
+}
+
+impl From<ProcedureIo> for Vec<u8> {
```

Comment:
The implementation for `From` for `ProcedureIO` is complemented with the automatic implementation of `Into` for the target type `Vec<u8>`, so no need here to implement it again :)

Reply:
But `Into<ProcedureIO>` is not implemented for `Vec<u8>` (and if it was, this `impl From<ProcedureIo> for Vec<u8>` would not be possible because of conflicting implementations). 
Or what exactly do you mean?

Reply:
yes, true that. I have overseen the tuple field inside [`ProcedureIo`]